pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go

```diff
@@ -683,13 +738,107 @@ func TestPodTopologySpreadScore(t *testing.T) {
+				st.MakePod().Namespace("default").Name("p-b1").Node("node-b").Label("bar", "").Obj(),
+				st.MakePod().Namespace("default").Name("p-b2").Node("node-b").Label("bar", "").Obj(),
+				st.MakePod().Namespace("default").Name("p-y1").Node("node-y").Label("foo", "").Obj(),
+				st.MakePod().Namespace("default").Name("p-y2").Node("node-y").Label("bar", "").Obj(),
+			},
+			config: &config.PodTopologySpreadArgs{
+				DefaultingType: config.ListDefaulting,
+				DefaultConstraints: []v1.TopologySpreadConstraint{
+					{MaxSkew: 3, TopologyKey: v1.LabelHostname, WhenUnsatisfiable: v1.ScheduleAnyway},
+				},
+			},
+			nodes: []*v1.Node{
+				st.MakeNode().Name("node-a").Label(v1.LabelTopologyZone, "zone1").Label(v1.LabelHostname, "node-a").Obj(),
+				st.MakeNode().Name("node-b").Label(v1.LabelTopologyZone, "zone1").Label(v1.LabelHostname, "node-b").Obj(),
+				st.MakeNode().Name("node-x").Label(v1.LabelHostname, "node-x").Obj(),
+				st.MakeNode().Name("node-y").Label(v1.LabelHostname, "node-y").Obj(),
+			},
+			failedNodes: []*v1.Node{},
+			objs: []runtime.Object{
+				&appsv1.ReplicaSet{ObjectMeta: metav1.ObjectMeta{Namespace: "default", Name: "rs1"}, Spec: appsv1.ReplicaSetSpec{Selector: st.MakeLabelSelector().Exists("bar").Obj()}},
+			},
+			want: []framework.NodeScore{
+				{Name: "node-a", Score: 100},
+				{Name: "node-b", Score: 40},
+				{Name: "node-x", Score: 100},
+				{Name: "node-y", Score: 80},
+			},
+		},
+		{
+			// +------------------------+------------------------+
+			// |         "zone1"        |          ""            |
+			// +------------------------+------------------------+
+			// | node-a |    node-b     | node-x |     node-y    |
+			// +--------+---------------+--------+---------------+
+			// | P{foo} | P{bar} P{bar} |        | P{foo} P{bar} |
+			// +--------+---------------+--------+---------------+
+			// For the first constraint (zone): the matching pods spread as  2/2/?1?/?1?
+			// For the second constraint (node): the matching pods spread as 0/2/0/1
+			name: "system default constraints",
+			pod:  st.MakePod().Name("p").Namespace("default").Label("bar", "").OwnerReference("rs1", appsv1.SchemeGroupVersion.WithKind("ReplicaSet")).Obj(),
+			existingPods: []*v1.Pod{
+				st.MakePod().Namespace("default").Name("p-a1").Node("node-a").Label("foo", "").Obj(),
+				st.MakePod().Namespace("default").Name("p-b1").Node("node-b").Label("bar", "").Obj(),
+				st.MakePod().Namespace("default").Name("p-b2").Node("node-b").Label("bar", "").Obj(),
+				st.MakePod().Namespace("default").Name("p-y1").Node("node-y").Label("foo", "").Obj(),
+				st.MakePod().Namespace("default").Name("p-y2").Node("node-y").Label("bar", "").Obj(),
+			},
+			config: &config.PodTopologySpreadArgs{
+				DefaultingType: config.SystemDefaulting,
+			},
+			nodes: []*v1.Node{
+				st.MakeNode().Name("node-a").Label(v1.LabelTopologyZone, "zone1").Label(v1.LabelHostname, "node-a").Obj(),
```

Comment:
We should focus on the case where all nodes don't have a zone label. If some of them do, we can't do better than considering them part of a single zone.

Unless your intent is that nodes that are not part of a zone behave like each of them is a separate zone?